### PR TITLE
feat: add pushed tags to metadata

### LINKED ISSUE #1341: list all pushed tags in the json metadata output of `oras push`
### What is the version of your ORAS CLI

1.2.0-beta.1

### What would you like to be added?

Currently all the pushed tags are displayed in the output but not in the json output of `oras push`
```console
$ oras push localhost:5000/test:tag1,tag2,tag3 --format json
✓ Uploaded  application/vnd.oci.empty.v1+json                                                                                    2/2  B 100.00%   48ms
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                                       535/535  B 100.00%   38ms
  └─ sha256:0a6d04eceb050df7446931a07d75d3465b28bd8616003eddcce6838cfdcd3995
Tagged tag3
Tagged tag2
{
  "Ref": "localhost:5000/test@sha256:0a6d04eceb050df7446931a07d75d3465b28bd8616003eddcce6838cfdcd3995",
  "MediaType": "application/vnd.oci.image.manifest.v1+json",
  "Digest": "sha256:0a6d04eceb050df7446931a07d75d3465b28bd8616003eddcce6838cfdcd3995",
  "Size": 535,
  "Annotations": {
    "org.opencontainers.image.created": "2024-04-12T08:26:14Z"
  },
  "ArtifactType": "application/vnd.unknown.artifact.v1"
}
```

### Why is this needed for ORAS?

Pushed tags `tag1` `tag2` and `tag3`  should be added to the  json output of `oras push` and selectable via go templating.

### Are you willing to submit PRs to contribute to this feature?

- [ ] Yes, I am willing to implement it.

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PRs adds pushed tags into the metadata output of `oras push`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1341 

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: cmd/oras/internal/display/metadata/model/push.go [hunk 1]
```diff
@@ -15,14 +15,25 @@ limitations under the License.
...
 }
 
 // NewPush returns a metadata getter for push command.
-func NewPush(desc ocispec.Descriptor, path string) any {
-	return push{FromDescriptor(path, desc)}
+func NewPush(desc ocispec.Descriptor, path string, tags []string) any {
+	var refAsTags []string
+	for _, tag := range tags {
+		refAsTags = append(refAsTags, path+":"+tag)
+
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/model/push.go:33]: nit: remove empty line

File: cmd/oras/internal/display/metadata/model/tag.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	tag.lock.Lock()
+	defer tag.lock.Unlock()
+	tag.tags = append(tag.tags, t)
+}
+
+// Tags returns the tags.
+func (tag *Tagged) Tags() []string {
+	tag.lock.RLock()
+	defer tag.lock.RUnlock()
+	return tag.tags
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/model/tag.go:39]: Do you want to sort it?

File: internal/contentutil/reference.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import "github.com/opencontainers/go-digest"
+
+// IsDigest checks if the given string is a valid digest.
+func IsDigest(tagOrDigest string) bool {
+	if _, err := digest.Parse(tagOrDigest); err == nil {
+		// is digest
+		return true
+	}
+	return false
```
Comment (shizhMSFT) [inline internal/contentutil/reference.go:26]: nit:
```suggestion
	_, err := digest.Parse(tagOrDigest)
	return err == nil
```

File: cmd/oras/internal/display/metadata/model/tag.go (hunk 2 above)
Comment (qweeah) [reply cmd/oras/internal/display/metadata/model/tag.go:47]: Added sorting during insertion

File: cmd/oras/internal/display/metadata/model/tag.go [hunk 4]
```diff
@@ -29,7 +29,15 @@ type Tagged struct {
...
 	defer tag.lock.Unlock()
+
 	tag.tags = append(tag.tags, t)
+	// sorted insert
+	for i := len(tag.tags) - 1; i > 0; i-- {
+		if tag.tags[i] >= tag.tags[i-1] {
+			break
+		}
+		tag.tags[i], tag.tags[i-1] = tag.tags[i-1], tag.tags[i]
+	}
```
Comment (shizhMSFT) [inline cmd/oras/internal/display/metadata/model/tag.go:40]: Consider [slice.Sort](https://pkg.go.dev/slices#Sort).

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**